### PR TITLE
formula: add slimmer api json v3 representation

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Alternative approach to https://github.com/Homebrew/brew/pull/16433 based on discussion there and in the original issue https://github.com/Homebrew/brew/issues/16410.

This is a second go at this after the recursive compacting approach in the earlier PR ended up not yielding big gains beyond what's in this PR. IMO this is the simplest approach to reducing the JSON size by omitting the bottle url field that we don't use anywhere and also removing blank fields at the top-level of the resulting hash.

I decided to memoize the variations generation since that takes a while but shouldn't change once a formula is loaded. This reduces time in the `brew generate-formula-api` command.

TODO:
- [x] Merge in https://github.com/Homebrew/brew/pull/16459
- [x] Add tests

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
Makes a lot of sense. I like the minimal approach here!

File: Library/Homebrew/test/formulary_spec.rb
Comment (apainintheneck) [inline Library/Homebrew/test/formulary_spec.rb:405]: By adding the `ruby_source_path` to the hash above and by mocking things so that it thinks it should use the API we no longer have to mock the `Formulary.loader_for` method. This means that the tests should better represent how the code is used in real life. It doesn't seem to make the tests noticeably slower either.

**Comment by apainintheneck** [comment]
As far as I can tell there should be no harm in merging this in as is. The `brew generate-formula-api` command will generate more files and take a little longer but those files will get ignored when the data is selectively archived and stored for later steps.

```yaml
- name: Archive data
  run: tar czvf data-core.tar.gz _data/formula/ _data/formula_canonical.json api/formula/ api/formula_tap_migrations.json formula/
```

From: https://github.com/Homebrew/formulae.brew.sh/blame/master/.github/workflows/scheduled.yml

I'm not entirely sure what changes need to be made in that repo after this ships to start shipping smaller JSON files to all users though.

**Comment by MikeMcQuaid** [comment]
@apainintheneck Happy to give final review and merge when no longer draft.

**Comment by apainintheneck** [comment]
I just wanted to open a WIP PR over on the `formulae.brew.sh` repo before merging this in to see if there were any changes I wanted to make (there was one).

- https://github.com/Homebrew/formulae.brew.sh/pull/1031

**Comment by apainintheneck** [comment]
Note that even with the memoization added to `Formula#to_hash_with_variations` I expect `brew generate-formula-api` to be 10-15 seconds slower than before. Of course, this is on an older machine but you get the idea.

### Before
```console
$ time brew generate-formula-api
________________________________________________________
Executed in  113.40 secs    fish           external
   usr time   96.75 secs  190.00 micros   96.75 secs
   sys time   13.92 secs  616.00 micros   13.92 secs
```

### After

```console
$ time brew generate-formula-api
________________________________________________________
Executed in  125.60 secs    fish           external
   usr time  103.79 secs  194.00 micros  103.79 secs
   sys time   15.89 secs  625.00 micros   15.89 secs
```

**Comment by apainintheneck** [comment]
This should probably not get merged in until the site generation is working again for `formulae.brew.sh`.

- https://github.com/Homebrew/formulae.brew.sh/issues/1030

**Comment by MikeMcQuaid** [comment]
> Note that even with the memoization added to `Formula#to_hash_with_variations` I expect `brew generate-formula-api` to be 10-15 seconds slower than before. Of course, this is on an older machine but you get the idea.

Seems reasonable, thanks @apainintheneck!

**Comment by Bo98** [comment]
Yes I was saying in the other issue that if we're doing a complete compatibility break here then now's the time to rethink things and start combining the files.

I also think redoing the list from an array to a `"name" => data` format will also help, basically incorporating at least part of this code into the format itself: 

https://github.com/Homebrew/brew/blob/34ff2946e7b96968ae8dc8c51228cb47c59b8e25/Library/Homebrew/api/formula.rb#L45-L56

https://github.com/Homebrew/brew/blob/34ff2946e7b96968ae8dc8c51228cb47c59b8e25/Library/Homebrew/api/cask.rb#L48-L57

**Comment by apainintheneck** [comment]
Yeah, I was planning on making the smallest possible change here that would benefit parsing performance. The changes here only affect file size not data layout. All of the other suggestions would change it's layout and possibly the download strategy which feels like something outside of the scope of this PR.

If you all still think that's the way to go, we need to decide on the new download strategy and file layouts before we can continue. Beyond that we'll have to update the API code so that it can manage both the old and new file formats temporarily. The slim cask JSON representation will need to be added in too.

**Comment by MikeMcQuaid** [comment]
@apainintheneck I think it's worth getting this into a stage/format that we can incrementally move to a better structure in future with minimal changes here and take this chance to reduce the number of downloads as well as the size. Feel free to push back on any of that, though, if you feel it blows up the scope too much.

File: Library/Homebrew/dev-cmd/generate-formula-api.rb
Comment (apainintheneck) [inline Library/Homebrew/dev-cmd/generate-formula-api.rb:63]: This is meant to match the formula cache in the `api/formula.rb` file.

https://github.com/Homebrew/brew/blob/36024e2a4db7127114c39bbdb62a812816e6a0bf/Library/Homebrew/api/formula.rb#L45-L47

**Comment by apainintheneck** [comment]
I think this is ready for final review. The main change here is that we're generating this new `internal_formula_v3.json` file which will end up getting generated automatically and hosted by `formulae.brew.sh`. Of course, it's a private API so it probably doesn't matter if it's available to download for a bit.

I'll make a follow-up PR to start signing those files in the `formulae.brew.sh` repo along with deleting the unsigned internal json files after I make a PR for the cask equivalent to this PR. Those should be trivial changes though.

Once we are generating both of the internal representation files, it will be much simpler to test using them to load packages locally. We can proceed with the environment variable approach of loading only for certain internal users and then testing it on larger groups of users. The final change to enable it for all users could wait for a minor release.

**Comment by Bo98** [comment]
Given we've now lifted up `renames` and `aliases`, do we need `oldnames` and `aliases` in the formulae hashes themselves anymore or can we strip them too?

**Comment by Bo98** [comment]
Would also be good to lift up tap info like tap name and tap_git_head.

**Comment by apainintheneck** [comment]
> Given we've now lifted up `renames` and `aliases`, do we need `oldnames` (and the deprecated `oldname`) and `aliases` in the formulae hashes themselves anymore or can we strip them too?

You're right I don't think those are really needed anymore. We can just skip the manual overrides in `Formulary.load_formula_from_api` and things should work correctly as long as we source the information correctly in the core tap.

#### To Skip
https://github.com/Homebrew/brew/blob/0ef5c28833f3d335eadd5e6e5d4956e9b15145e9/Library/Homebrew/formulary.rb#L380-L388

#### To Update
https://github.com/Homebrew/brew/blob/13c4d0ac11fbe4af23339183ef7ae156b8a10b6d/Library/Homebrew/tap.rb#L686-L696

---

I had this idea that we were caching the formula source file but it looks like we only do that for casks and that's why I got confused. If we're caching the json source, then we do need all of this information to be there when installing since we won't have the whole API JSON file at uninstall time. Of course, we could just add that info manually too.

https://github.com/Homebrew/brew/blob/13c4d0ac11fbe4af23339183ef7ae156b8a10b6d/Library/Homebrew/cask/cask_loader.rb#L257

**Comment by apainintheneck** [comment]
> Would also be good to lift up tap info like tap name and tap_git_head.

I don't think tap name gets used at all though but git head wouldn't be bad to add. Then, we can just reference it in `CoreTap` if we need to access `tap_git_head`.

#### To Skip
https://github.com/Homebrew/brew/blob/13c4d0ac11fbe4af23339183ef7ae156b8a10b6d/Library/Homebrew/formulary.rb#L375-L378

**Comment by apainintheneck** [comment]
Of course, the changes to formula loading will happen in a follow-up PR. I'll just omit more fields from the JSON based on this.

**Comment by apainintheneck** [comment]
Just to be double sure, we always store the Ruby source file in something like `$(brew --cellar)/foo/1.0/.brew/foo.rb` so we don't have to worry about backwards compatibility with loading old cached JSON representations of installed formulae. Is that right?

It seems like that's only going to be a concern for casks.

File: Library/Homebrew/formula.rb
Comment (apainintheneck) [inline Library/Homebrew/formula.rb:2456]: https://github.com/Homebrew/brew/blob/13c4d0ac11fbe4af23339183ef7ae156b8a10b6d/Library/Homebrew/formula.rb#L219-L220

**Comment by apainintheneck** [comment]
Okay, I omitted a few more fields from the hash that were already included at the top-level. I also removed a few values that default to zero if they are in fact zero.

- https://github.com/Homebrew/brew/pull/16460/commits/ba09df4c396db0d0e5abaf7ee986c9e1acb3fdde

File: Library/Homebrew/dev-cmd/generate-formula-api.rb
Comment (Bo98) [inline Library/Homebrew/dev-cmd/generate-formula-api.rb:60]:
```suggestion
        "tap"            => {
          "name"     => tap.name,
          "git_head" => tap.git_head,
        },
```

maybe?

**Review by Bo98 — APPROVED** [review]
Looking good

**Comment by Bo98** [comment]
> Just to be double sure, we always store the Ruby source file in something like `$(brew --cellar)/foo/1.0/.brew/foo.rb` so we don't have to worry about backwards compatibility with loading old cached JSON representations of installed formulae.

Yes, that's correct.

File: Library/Homebrew/dev-cmd/generate-formula-api.rb
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/generate-formula-api.rb:60]: I went back and forth on that one a bit. Technically all of this information relates to a tap though so it seems kind of like unnecessary nesting.

Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/generate-formula-api.rb:60]: Maybe we should just name the file `formula_tap.json` and remove the tap prefix from all of the hash keys?

Comment (Bo98) [reply Library/Homebrew/dev-cmd/generate-formula-api.rb:60]: Ok, I can see that argument that's fair

Comment (Bo98) [reply Library/Homebrew/dev-cmd/generate-formula-api.rb:60]: > Maybe we should just name the file formula_tap.json

If anything it could be in reflection of the tap name so we have in theory homebrew-core.json, homebrew-cask.json and a (purely as an example) homebrew-cask-fonts.json would slot in nicely. Then there's probably no need to store the name in the JSON itself.

Comment (Bo98) [reply Library/Homebrew/dev-cmd/generate-formula-api.rb:60]: > I don't think we even use the tap name when initializing formulas from the API so it might not be necessary to include anyway.

Probably not but if we move it to live in the tap class then it will probably be easier that way rather than saying "homebrew-core" = "formula" and "homebrew-cask" = "cask".

Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/generate-formula-api.rb]: Handled in https://github.com/Homebrew/brew/pull/16460/commits/bc029cfca31386bdb3764619956bac8d1fab31f0

**Comment by MikeMcQuaid** [comment]
> I don't think tap name gets used at all

Let's not have anything (or as little as possible) in the internal API that we don't use. YAGNI and all that.

> Just to be double sure, we always store the Ruby source file in something like `$(brew --cellar)/foo/1.0/.brew/foo.rb` so we don't have to worry about backwards compatibility with loading old cached JSON representations of installed formulae.

Yes, at least for anything other than truly ancient kegs.

File: Library/Homebrew/dev-cmd/generate-formula-api.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/generate-formula-api.rb:88]:
```suggestion
      File.write("api/homebrew-core-internal.json", homebrew_core_tap_json) unless args.dry_run?
```

or something, unless I've missed something here.

File: Library/Homebrew/formula.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:2460]: Could also probably remove the `root_url` given it's the same for every bottle in the tap.

File: Library/Homebrew/formula.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:2451]: Could maybe also do this for keys that are `null` (if you haven't already) e.g. `urls.stable.using` for most formulae

File: Library/Homebrew/test/formulary_spec.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/test/formulary_spec.rb:409]: Very nice 🎉

File: Library/Homebrew/formula.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:2448]: - Could remove `options`, nothing in homebrew/core has them.
- Can remove `linked_keg`, `installed`, `pinned`, `outdated` as it's a reflection of the machine that generated this output

**Review by MikeMcQuaid — APPROVED** [review]
Only the filename change (unless I've missed something) is blocking. The rest are all optional and you can self-merge whenever you feel comfortable. Thanks!

File: Library/Homebrew/formula.rb
Comment (apainintheneck) [reply Library/Homebrew/formula.rb:2448]: I believe those will all be omitted automatically by the `api_has.reject { |_, value| value.blank? }` clause later on.

File: Library/Homebrew/dev-cmd/generate-formula-api.rb
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/generate-formula-api.rb:88]: It was an idea from @Bo98 that I liked. See this thread starting here https://github.com/Homebrew/brew/pull/16460#discussion_r1460722950

File: Library/Homebrew/formula.rb
Comment (apainintheneck) [reply Library/Homebrew/formula.rb:2451]: There were diminishing returns with trying to remove nested blank fields in the previous PR but I can just double-check that right now.

File: Library/Homebrew/formula.rb
Comment (apainintheneck) [reply Library/Homebrew/formula.rb:2460]: Yeah, that could probably just be at the top-level.

File: Library/Homebrew/dev-cmd/generate-formula-api.rb
Comment (Bo98) [reply Library/Homebrew/dev-cmd/generate-formula-api.rb:88]: Might make sense to lean on path components so it's `/api/internal/v3/homebrew-core.json`. The version being a "directory" like that is very common practice.

File: Library/Homebrew/formula.rb
Comment (apainintheneck) [reply Library/Homebrew/formula.rb:2460]: Actually, it's not used at all.

https://github.com/Homebrew/brew/blob/86e1c8aacf6196e10cef705fe01fc8d8f6737ec4/Library/Homebrew/formulary.rb#L293-L297

File: Library/Homebrew/dev-cmd/generate-formula-api.rb
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/generate-formula-api.rb:88]: The second hardest problem is CS strikes again. I'm fine with that proposal.

File: Library/Homebrew/formula.rb
Comment (Bo98) [reply Library/Homebrew/formula.rb:2468]: Yeah it was removed for security reasons prior to the introduction of JWS. Those reasons aren't really relevant anymore but equally the API should be able to behave like `brew bottle` and omit `root_url` if it is the default.

Even that code that's there now on the Formulary feels unnecessary - surely we can omit it like we omit it for Ruby formulae.

**Review by MikeMcQuaid — APPROVED** [review]
A few more comments but I'd be fine with merging as-is and iterating once merged whenever you prefer that. Great work!

File: Library/Homebrew/test/formulary_spec.rb
Comment (apainintheneck) [reply Library/Homebrew/test/formulary_spec.rb:517]: Honestly, real tests will have to wait until we're actually building this on CI. I'm guessing that some other tweaks will be necessary to get things working afterwards anyway.

File: Library/Homebrew/formula.rb
Comment (apainintheneck) [reply Library/Homebrew/formula.rb:2448]: Is it time for v3 of this PR? 😆

I prefer the idea of explicitly building the hash over the shared constant. That would require a new `#to_api_hash_with_variations` along with building the hash explicitly in `#to_api_hash`. As a consequence of that the runtime of the `brew generate-*-api` methods will probably be double what it is currently since we're doing twice the work. Building the variations is especially slow. If that sounds good, I'll close this PR and continue with that idea. Hopefully we won't need a v4.

Comment (apainintheneck) [reply Library/Homebrew/formula.rb:2448]: > `linked_keg`, `installed`, `pinned`, `outdated` will not be removed depending on the values on the generated machine (which seems undesirable).

But we know which machines we're running these commands on so that shouldn't be a problem, right?

File: Library/Homebrew/dev-cmd/generate-formula-api.rb
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/generate-formula-api.rb:88]: Note to self that this requires creating some new directories too (think `mkdir -p`).

**Comment by apainintheneck** [comment]
Closing this out in favor of the more explicit approach detailed in this thread: https://github.com/Homebrew/brew/pull/16460#discussion_r1461844684

File: Library/Homebrew/test/formulary_spec.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/test/formulary_spec.rb:517]: @apainintheneck Yeh, sorry, I guess what I meant here is that the best assertion here would be that `expect(formula.to_api_hash).to eql { ... }` with some actual hash where we can compare the output.

**Pull request closed without merging**
